pkg/controller/deployment/deployment_controller.go

```diff
@@ -438,41 +439,146 @@ func (dc *DeploymentController) syncDeployment(key string) error {
+// when a deployment is paused and not during the normal rollout process.
+func (dc *DeploymentController) scale(deployment *extensions.Deployment, newRS *extensions.ReplicaSet, oldRSs []*extensions.ReplicaSet) error {
+	// Paused and recreate deployments may not have a new replica set. Additionally,
+	// in case no old replica sets have pods, the controller should scale the new
+	// replica set to the full count.
+	if newRS != nil && len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+		if newRS.Spec.Replicas == deployment.Spec.Replicas {
+			return nil
+		}
+		if _, _, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// If the new replica set is saturated, old replica sets should be fully scaled down.
+	// This case handles replica set adoption during a saturated new replica set.
+	if deploymentutil.IsSaturated(deployment, newRS) {
+		for _, old := range controller.FilterActiveReplicaSets(oldRSs) {
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(old, 0, deployment); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// There are old replica sets with pods and the new replica set is not saturated.
+	// We need to proportionally scale all replica sets (new and old) in case of a
+	// rolling deployment.
+	if deploymentutil.IsRollingUpdate(deployment) {
+		allRSs := controller.FilterActiveReplicaSets(append(oldRSs, newRS))
+		allRSsReplicas := deploymentutil.GetReplicaCountForReplicaSets(allRSs)
+
+		// Number of additional replicas that can be either added or removed from the total
+		// replicas count. These replicas should be distributed proportionally to the active
+		// replica sets. Any leftovers should be distributed as well.
+		deploymentReplicasToAdd := deployment.Spec.Replicas - allRSsReplicas
+		if deploymentReplicasToAdd == 0 {
+			return nil
+		}
+
+		// The additional replicas should be distributed proportionally amongst the active
+		// replica sets from the larger to the smaller in size replica set. Scaling direction
+		// drives what happens in case we are trying to scale replica sets of the same size.
+		// In such a case when scaling up, we should scale up newer replica sets first, and
+		// when scaling down, we should scale older replica sets first.
+		scaleUp := deploymentReplicasToAdd > 0
+		scalingOperation := "up"
+		if scaleUp {
+			sort.Sort(controller.ReplicaSetsBySizeNewer(allRSs))
+		} else {
+			sort.Sort(controller.ReplicaSetsBySizeOlder(allRSs))
+			scalingOperation = "down"
+		}
+
+		// Iterate over all active replica sets and estimate proportions for each of them.
+		// The absolute value of deploymentReplicasAdded should never exceed the absolute
+		// value of deploymentReplicasToAdd.
+		deploymentReplicasAdded := int32(0)
+		for i := range allRSs {
+			rs := allRSs[i]
+
+			proportion := getProportion(rs, allRSsReplicas, deploymentReplicasToAdd, deploymentReplicasAdded)
```

Comment:
This is true only when we scale up but I agree. Added the check in the update loop below.